src/jit/lowerxarch.cpp

```diff
@@ -1030,6 +1034,31 @@ void Lowering::TreeNodeInfoInitShiftRotate(GenTree* tree)
     GenTreePtr shiftBy = tree->gtOp.gtOp2;
     GenTreePtr source  = tree->gtOp.gtOp1;
 
+#ifdef _TARGET_X86_
+    // The first operand of a GT_LSH_HI and GT_RSH_LO oper is a GT_LONG so that
+    // we can have a three operand form. Increment the srcCount.
+    if (tree->OperGet() == GT_LSH_HI || tree->OperGet() == GT_RSH_LO)
+    {
+        assert(source->OperGet() == GT_LONG);
+
+        info->srcCount++;
```

Comment:
Something happened to remember signing off:

For codgen we don't need sourceHi to be always in reg in case of GT_LSH_HI (because it could be moved from memory to targetReg if sourceHi is a contained mem-op).  Similarly for GT_RSH_LO, sourceLo could be marked as contained memory-op.  Even if not a memory-op, we could mark it as reg-optional.

It is ok not to implement contained mem-op/reg-optional in the first implementation.  But please capture as a TODO-CQ comment and an issue, so that we can revisit this in future.